### ISSUE TITLE
Problems with training parameter settings

### ISSUE THREAD

**Issue by zyy-xy** [ultralytics/ultralytics#12725]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello! Excuse me,
1. I'm working on a training task with specific 3 classes that are not in the 80 classes of the COCO dataset. Now I want to retrain on the official model, which may require the `freeze` parameter, but I don't know how to set it up. After researching, I found that I may need to set an integer value to represent the number of layers frozen, but I don't know what that layer is, where can I see the specific structure code of the model or the document that explains what each layer is? In my project, how many layers of training do you recommend to freeze?
2. If I need to modify the network structure, is it effective to modify the `yolov8.yaml` file directly?

Could you please help me?

**Comment by zyy-xy** [comment]
Also, I want to use my own data augmentation function, does the 'auto_augment' parameter meet my needs? How should this parameter be used? Are there any requirements for the input and output of data augmentation functions?

Let's say I have a data augmentation function `transfer_image`, how can I use it in my actual training? Can you show me an example?

**Comment by zyy-xy** [comment]
Also, I want to use my own data augmentation function, does the `auto_augment` parameter meet my needs? How should this parameter be used? Are there any requirements for the input and output of data augmentation functions?

Let's say I have a data augmentation function `transfer_image`, how can I use it in my actual training? Can you show me an example?

**Comment by glenn-jocher** [comment]
Hello!

The `auto_augment` option provides built-in support for data augmentation. However, if you want to implement custom augmentation methods like your `transfer_image` function, you would need to manually modify the data loading and augmentation code within the training loop, as it can't be integrated directly through YAML files or parameters provided by the CLI.

Here's a basic example on where you might integrate your `transfer_image` function:

```python
from ultralytics import YOLO

# Define your custom augmentation function
def transfer_image(image, labels):
    # Apply your transformations
    # Make sure to adjust labels as needed if transformations affect them
    return image, labels

# Load a model
model = YOLO('yolov8n.yaml')

# Customize the dataset with your augmentation
class CustomDataset(model.datasets['train']):
    def __getitem__(self, index):
        img, labels = super().__getitem__(index)
        return transfer_image(img, labels)

# Replace the original dataset with your custom one
model.datasets['train'] = CustomDataset

# Train the model
results = model.train()
```

This code is meant to point you in the right direction, as implementing it into your pipeline will depend on the specifics of your augmentation functions and how they interact with image and label data. Remember to maintain the data integrity, especially the bounding box or annotations after augmentation. Happy coding! 🚀

**Comment by zyy-xy** [comment]
> Hello!
> 
> The option provides built-in support for data augmentation. However, if you want to implement custom augmentation methods like your function, you would need to manually modify the data loading and augmentation code within the training loop, as it can't be integrated directly through YAML files or parameters provided by the CLI.`auto_augment``transfer_image`
> 
> Here's a basic example on where you might integrate your function:`transfer_image`
> 
> ```python
> from ultralytics import YOLO
> 
> # Define your custom augmentation function
> def transfer_image(image, labels):
>     # Apply your transformations
>     # Make sure to adjust labels as needed if transformations affect them
>     return image, labels
> 
> # Load a model
> model = YOLO('yolov8n.yaml')
> 
> # Customize the dataset with your augmentation
> class CustomDataset(model.datasets['train']):
>     def __getitem__(self, index):
>         img, labels = super().__getitem__(index)
>         return transfer_image(img, labels)
> 
> # Replace the original dataset with your custom one
> model.datasets['train'] = CustomDataset
> 
> # Train the model
> results = model.train()
> ```
> 
> This code is meant to point you in the right direction, as implementing it into your pipeline will depend on the specifics of your augmentation functions and how they interact with image and label data. Remember to maintain the data integrity, especially the bounding box or annotations after augmentation. Happy coding! 🚀

Thanks! @glenn-jocher And I have 2 more questions.
1. I'm working on a training task with specific 3 classes that are not in the 80 classes of the COCO dataset. Now I want to retrain on the official model, which may require the parameter `freeze`, but I don't know how to set it up. After researching, I found that I may need to set an integer value to represent the number of layers frozen, but I don't know what that layer is, where can I see the specific structure code of the model or the document that explains what each layer is? In my project, how many layers of training do you recommend to freeze?
2. If I need to modify the network structure, is it effective to modify the file `yolov8.yaml` directly?

**Comment by glenn-jocher** [comment]
Hello!

Great questions!

1. **Freezing Layers**: To decide how many layers to freeze, you typically start by freezing early layers which capture basic features like edges and textures. For YOLO models, a common approach is to freeze the backbone or parts of it. You can view the layer structure directly in the model definition file (`yolov8n.yaml` or similar). Start by freezing the first half of layers and adjust based on your fine-tuning results. Here is an example of how to set the `freeze` parameter:

    ```bash
    yolo train data=mydata.yaml model=yolov8n.pt freeze=10
    ```
   This would freeze the first 10 layers. Experiment with this number to find a balance between adapting to your specific tasks and leveraging learned features.

2. **Modifying Network Structure**: Yes, modifying the `yolov8.yaml` file is effective and recommended for structural changes in your network. Make sure to keep backups of original configurations and test incrementally to understand the impact of any changes.

Happy training and fine-tuning! 🚀

**Comment by glenn-jocher** [comment]
Hello!

Currently, YOLOv8 does not support different NMS thresholds for different categories directly through the CLI or standard configuration files. However, you can achieve this by modifying the source code where the NMS function is called.

Here's a brief example of how you might adjust the NMS threshold per class in Python:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.pt')

# Define custom NMS thresholds per class
class_thresholds = {0: 0.3, 1: 0.4, 2: 0.5}  # Example thresholds for three classes

# Modify the NMS process
def custom_nms(boxes, class_thresholds):
    # Apply per-class NMS threshold
    for box in boxes:
        cls = box.cls  # Assuming 'cls' is the class index
        box.conf = box.conf if box.conf > class_thresholds.get(cls, 0.25) else 0
    return boxes

# Apply during prediction
results = model.predict('path/to/image.jpg')
filtered_boxes = custom_nms(results.boxes, class_thresholds)
```

This is a simplified example. You'll need to integrate it properly depending on how the NMS is implemented in your version of YOLOv8.

Hope this helps! 🚀

**Comment by zyy-xy** [comment]
> Hello!
> 
> The `auto_augment` option provides built-in support for data augmentation. However, if you want to implement custom augmentation methods like your `transfer_image` function, you would need to manually modify the data loading and augmentation code within the training loop, as it can't be integrated directly through YAML files or parameters provided by the CLI.
> 
> Here's a basic example on where you might integrate your `transfer_image` function:
> 
> ```python
> from ultralytics import YOLO
> 
> # Define your custom augmentation function
> def transfer_image(image, labels):
>     # Apply your transformations
>     # Make sure to adjust labels as needed if transformations affect them
>     return image, labels
> 
> # Load a model
> model = YOLO('yolov8n.yaml')
> 
> # Customize the dataset with your augmentation
> class CustomDataset(model.datasets['train']):
>     def __getitem__(self, index):
>         img, labels = super().__getitem__(index)
>         return transfer_image(img, labels)
> 
> # Replace the original dataset with your custom one
> model.datasets['train'] = CustomDataset
> 
> # Train the model
> results = model.train()
> ```
> 
> This code is meant to point you in the right direction, as implementing it into your pipeline will depend on the specifics of your augmentation functions and how they interact with image and label data. Remember to maintain the data integrity, especially the bounding box or annotations after augmentation. Happy coding! 🚀

Thanks! @glenn-jocher  And I have another question. I want to transform the image with a custom function at a specified probability during training, just like any other data augmentation option, what should I do? (Assuming my annotations don't change when the image is transformed)

**Comment by glenn-jocher** [comment]
@zyy-xy hello!

To integrate a custom augmentation function with a specified probability during training, you can modify the dataset class to apply your function conditionally. Here's an example:

```python
from ultralytics import YOLO
import random

# Define your custom augmentation function
def transfer_image(image, labels):
    # Apply your transformations
    return image, labels

# Load a model
model = YOLO('yolov8n.yaml')

# Customize the dataset with your augmentation
class CustomDataset(model.datasets['train'].__class__):
    def __getitem__(self, index):
        img, labels = super().__getitem__(index)
        if random.random() < 0.5:  # 50% probability
            img, labels = transfer_image(img, labels)
        return img, labels

# Replace the original dataset with your custom one
model.datasets['train'] = CustomDataset(model.datasets['train'].data, model.datasets['train'].labels)

# Train the model
results = model.train()
```

This example applies your `transfer_image` function with a 50% probability. Adjust the probability as needed. Happy training! 🚀
